rmw_zenoh_cpp/src/detail/liveliness_utils.cpp

```diff
@@ -232,19 +264,26 @@ std::optional<rmw_qos_profile_t> keyexpr_to_qos(const std::string & keyexpr)
   }
 
   try {
-    qos.history = str_to_qos_history.at(history_parts[0]);
-    qos.reliability = str_to_qos_reliability.at(parts[0]);
-    qos.durability = str_to_qos_durability.at(parts[1]);
-    qos.liveliness = str_to_qos_liveliness.at(liveliness_parts[0]);
+    qos.history = history_parts[0].empty() ? default_qos.history : str_to_qos_history.at(
+      history_parts[0]);
+    qos.reliability = parts[0].empty() ? default_qos.reliability : str_to_qos_reliability.at(
+      parts[0]);
+    qos.durability = parts[1].empty() ? default_qos.durability : str_to_qos_durability.at(parts[1]);
+    qos.liveliness =
+      liveliness_parts[0].empty() ? default_qos.liveliness : str_to_qos_liveliness.at(
+      liveliness_parts[0]);
   } catch (const std::exception & e) {
     RMW_SET_ERROR_MSG_WITH_FORMAT_STRING("Error setting QoS values from strings: %s", e.what());
     return std::nullopt;
   }
 
   // Helper function to convert string to size_t.
   auto str_to_size_t =
-    [](const std::string & str) -> std::optional<size_t>
+    [](const std::string & str, const std::size_t default_value) -> std::optional<size_t>
```

Comment:
I'm concerned about the `size_t` here.  In particular, it is not always the case that `sizeof(size_t)` == `sizeof(uint64_t)`.  Yet the deadline, lifespan, and liveliness changes down below assume that it is the case (for the `sec` and `nsec`).

What I'm going to suggest instead is that we make this a "real" function, and templatize `default_value`.  Then all of those will get code generated properly.

Something like:

```
template<typename T>
static std::optional<T> str_to_size_t(const std::string & str, const T default_value)
{
  if (str.empty()) {
    return default_value;
  }
  errno = 0;
  char * endptr;
  size_t num = strtoul(str.c_str(), &endptr, 10);
  if (endptr == str.c_str()) {
    // No values were converted, this is an error
    RMW_SET_ERROR_MSG("no valid numbers available");
    return std::nullopt;
  } else if (*endptr != '\0') {
    // There was junk after the number
    RMW_SET_ERROR_MSG("non-numeric values");
    return std::nullopt;
  } else if (errno != 0) {
    // Some other error occurred, which may include overflow or underflow
    RMW_SET_ERROR_MSG(
      "an undefined error occurred while getting the number, this may be an overflow");
    return std::nullopt;
  }
  return num;
}
```

Reply:
Great suggestion; I've modified accordingly https://github.com/ros2/rmw_zenoh/pull/266/commits/aa5697d9c0c5bb8e8da7c4a743048fe7324734dc